ivre/db/mongo.py

```diff
@@ -1142,6 +1142,32 @@ def searchscreenshot(port=None, protocol='tcp', service=None, neg=False):
                            'screenshot': {'$exists': not neg}}
         }}
 
+    @staticmethod
+    def searchcpe(value=None, type=None, vendor=None, product=None,
+                  components=None):
+        """Look for a CPE by value (original cpe string), type (a, o or h),
+        vendor, product or components (the part after the column following
+        the product). No argument will just check for cpe existence.
+
+        """
+        if all(arg is None
+               for arg in [value, type, vendor, product, components]):
+            return {"cpes": {"$exists": True}}
+
+        flt = {}
+        fields = [
+            ("value", value),
+            ("type", type),
+            ("vendor", vendor),
+            ("product", product),
+            ("components", components),
+        ]
+
+        for (field_name, field_val) in fields:
+            if field_val is not None:
+                flt["cpes." + field_name] = field_val
+        return flt
```

Comment:
First of all, I think CPE indexes should be [sparse](http://docs.mongodb.org/manual/core/index-sparse/).

I think at least an index on `value` is mandatory. Then, I'm not sure. We can either opt for
  - an index per field.
  - a [compound index](http://docs.mongodb.org/manual/core/index-compound/) that encompasses all the (other) fields in the correct order (this would be coherent with the choices made for `service`, `product`, and `version`).
  - no other index, and build the `value` to query (or at least a regular expression) based on the provided fields.

In any case, I think we should try, in `searchcpe()`, to complete the query. For example, if the user provides `product="lighttpd"`, we might want to guess `vendor="lighttpd", type="a"` and possibly `value=re.compile("^cpe:/a:lighttpd:lighttpd:")`.

This is particularly useful if we want to go for option 2 or 3 for indexing.

What do you think?